connector/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisRecordProcessor.scala

```diff
@@ -91,56 +88,68 @@ private[kinesis] class KinesisRecordProcessor[T](receiver: KinesisReceiver[T], w
            *  This will potentially cause records since the last checkpoint to be processed
            *     more than once.
            */
-          logError(log"Exception: WorkerId ${MDC(WORKER_URL, workerId)} encountered and " +
-            log"exception while storing or checkpointing a batch for workerId " +
-            log"${MDC(WORKER_URL, workerId)} and shardId ${MDC(SHARD_ID, shardId)}.", e)
+          logError(log"Exception: SchedulerId ${MDC(WORKER_URL, schedulerId)} encountered and " +
+            log"exception while storing or checkpointing a batch for schedulerId " +
+            log"${MDC(WORKER_URL, schedulerId)} and shardId ${MDC(SHARD_ID, shardId)}.", e)
 
-          /* Rethrow the exception to the Kinesis Worker that is managing this RecordProcessor. */
+          /* Rethrow the exception to the Kinesis scheduler that is managing
+           this RecordProcessor. */
           throw e
       }
     } else {
       /* RecordProcessor has been stopped. */
-      logInfo(log"Stopped: KinesisReceiver has stopped for workerId ${MDC(WORKER_URL, workerId)}" +
-          log" and shardId ${MDC(SHARD_ID, shardId)}. No more records will be processed.")
+      logInfo(log"Stopped: KinesisReceiver has stopped for schedulerId " +
+        log"${MDC(WORKER_URL, schedulerId)} and shardId ${MDC(SHARD_ID, shardId)}. " +
+        log"No more records will be processed.")
     }
   }
 
   /**
-   * Kinesis Client Library is shutting down this Worker for 1 of 2 reasons:
-   * 1) the stream is resharding by splitting or merging adjacent shards
-   *     (ShutdownReason.TERMINATE)
-   * 2) the failed or latent Worker has stopped sending heartbeats for whatever reason
-   *     (ShutdownReason.ZOMBIE)
+   * Called when the lease that tied to this Kinesis record processor has been lost.
+   * Once the lease has been lost the record processor can no longer checkpoint.
+   *
+   * @param leaseLostInput gives access to information related to the loss of the lease.
+   *   Currently this has no functionality.
+   */
+  override def leaseLost(leaseLostInput: LeaseLostInput): Unit = {
+    logInfo(log"The lease for shardId: ${MDC(SHARD_ID, shardId)} is lost.")
+    receiver.removeCheckpointer(shardId, null)
+  }
+
+  /**
+   * Called when the shard that this Kinesis record processor is handling has been completed.
+   * Once a shard has been completed no further records will ever arrive on that shard.
    *
-   * @param checkpointer used to perform a Kinesis checkpoint for ShutdownReason.TERMINATE
-   * @param reason for shutdown (ShutdownReason.TERMINATE or ShutdownReason.ZOMBIE)
+   * When this is called the record processor <b>must</b> checkpoint. Otherwise an exception
+   * will be thrown and the all child shards of this shard will not make progress.
+   *
+   * @param shardEndedInput provides access to a checkpointer method for completing processing of
+   *   the shard.
    */
-  override def shutdown(
```

Comment:
`shutdown` is divided into three functions for the corresponding shutdown reasons.
* TERMINATE => shardEnded
* REQUESTED => shutdownRequested
* ZOMBIE => leaseLost

In the current code, `REQUESTED` is not cared but in this case checkpoint can be done. So, checkpoint is done in `shutdownRequested`.